### PR TITLE
cxxtest: init at 4.4

### PR DESCRIPTION
**Pull request by juliendehos**
###### Things done
- [ ] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [X] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

NB: wrapPython changes the interpreter (from python to sh) for the main script (cxxtestgen), so I had to update the cmake module too.

---

### CODE REVIEW

File: pkgs/development/tools/build-managers/cmake/default.nix [hunk 1]
```diff
@@ -28,7 +28,8 @@ stdenv.mkDerivation rec {
   patches =
     # Don't search in non-Nix locations such as /usr, but do search in
     # Nixpkgs' Glibc.
-    optional (stdenv ? glibc) ./search-path-3.2.patch
+    [ ./cxxtest_cmake.patch ]
+    ++ optional (stdenv ? glibc) ./search-path-3.2.patch
```
Comment (rasendubi) [inline pkgs/development/tools/build-managers/cmake/default.nix]: The comment above is not off.

File: pkgs/development/tools/build-managers/cmake/cxxtest_cmake.patch [hunk 2]
```diff
@@ -0,0 +1,11 @@
+--- a/Modules/FindCxxTest.cmake	2016-01-25 17:57:19.000000000 +0100
```
Comment (rasendubi) [inline pkgs/development/tools/build-managers/cmake/cxxtest_cmake.patch]: Any comment why the patch is needed? (It looks like a hack to me.)

File: pkgs/development/libraries/cxxtest/default.nix [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+  installPhase = ''
+    cd python
+    python setup.py install --prefix=$out
+    cd ..
+
+    mkdir -p $out/include
+    cp -R cxxtest $out/include/
+
+    wrapProgram $out/bin/cxxtestgen --prefix PYTHONPATH : "$(toPythonPath $out)"
```
Comment (rasendubi) [inline pkgs/development/libraries/cxxtest/default.nix]: Maybe use `wrapPython` instead.

**Comment by juliendehos** [comment]
@rasendubi: cxxtest is a python module + a python script (cxxtestgen). Using wrapPython, wraps cxxtestgen in a shell script but cmake config for cxxtest still calls cxxtestgen using the python interpreter, hence the patch. I know this patch is a hack; maybe I can wrap the shell wrap in another python script and cancel the patch if you think it's a better solution.

File: pkgs/development/libraries/cxxtest/default.nix [hunk 4]
```diff
@@ -0,0 +1,35 @@
+{ stdenv, fetchFromGitHub, python, pythonPackages}:
+
```
Comment (FRidh) [inline pkgs/development/libraries/cxxtest/default.nix]: note that python is an attribute of pythonPackages so you can drop python here

File: pkgs/development/libraries/cxxtest/default.nix [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    repo = name;
+    rev = version;
+    sha256 = "19w92kipfhp5wvs47l0qpibn3x49sbmvkk91yxw6nwk6fafcdl17";
+  };
+
+  buildInputs = with pythonPackages; [ python wrapPython ];
+
+  installPhase = ''
+    cd python
+    python setup.py install --prefix=$out
```
Comment (FRidh) [inline pkgs/development/libraries/cxxtest/default.nix]: it seems that there is a `setup.py` file. Any specific reason for not using `pythonPackages.buildPythonPackage` ?

File: pkgs/development/tools/build-managers/cmake/default.nix [hunk 6]
```diff
@@ -25,10 +25,11 @@ stdenv.mkDerivation rec {
     sha256 = "1yl0z422gr7zfc638chifv343vx0ig5gasvrh7nzf7b15488qgxp";
   };
 
-  patches =
-    # Don't search in non-Nix locations such as /usr, but do search in
-    # Nixpkgs' Glibc.
-    optional (stdenv ? glibc) ./search-path-3.2.patch
```
Comment (FRidh) [inline pkgs/development/tools/build-managers/cmake/default.nix]: this needs to be a separate commit since you touch another package.

File: pkgs/development/libraries/cxxtest/default.nix (hunk 5 above)
Comment (juliendehos) [inline pkgs/development/libraries/cxxtest/default.nix]: The python package is used in the main script "cxxtestgen". This package is very specific so I don't think it can be useful in another application.

File: pkgs/development/tools/build-managers/cmake/default.nix (hunk 6 above)
Comment (juliendehos) [inline pkgs/development/tools/build-managers/cmake/default.nix]: Done: #17426.

File: pkgs/development/libraries/cxxtest/default.nix (hunk 5 above)
Comment (FRidh) [inline pkgs/development/libraries/cxxtest/default.nix]: I don't mean like adding it to say `python-packages.nix` but more that `buildPythonPackage` takes care of a lot of things

Comment (juliendehos) [inline pkgs/development/libraries/cxxtest/default.nix]: Oops, sorry.

Indeed, using `buildPythonPackage` I can remove the line 
```
wrapPythonProgramsIn $out/bin "$out $pythonPath"
```
but I have to add the lines
```
  buildPhase = ''
    pythonInstallDir=$out/lib/${pythonPackages.python.libPrefix}/site-packages
    mkdir -p $pythonInstallDir
    export PYTHONPATH=''${PYTHONPATH:+''${PYTHONPATH}:}$pythonInstallDir
    cd python
    python setup.py build
    cd ..
  '';
  doCheck = false;
```
so the first version seems simpler to me.

**Comment by FRidh** [comment]
This looks good to me then. Let's wait on #17426.

**Comment by FRidh** [comment]
If you rebase this and cc me then it should be merged in time before we branch of the new stable

**Comment by juliendehos** [comment]
I've rebased, compiled and tested. It seems ok.

File: pkgs/development/libraries/cxxtest/default.nix (hunk 5 above)
Comment (FRidh) [inline pkgs/development/libraries/cxxtest/default.nix]: one more comment for here though. You wrote `python setup.py`. Typically we use `${python.interpreter}` so it will also work when `python3` is used

Comment (juliendehos) [inline pkgs/development/libraries/cxxtest/default.nix]: Yes. I've updated the commit.
